### PR TITLE
search: Fix getting results for non-existant files

### LINKED ISSUE #17907: Don't search files that don't exist on certain branches/commits on structural search path
Seeing the `git ls-tree ... file does not exist` first seen in https://github.com/sourcegraph/sourcegraph/pull/17726#issue-562835104, but now in the wild, with this query example [`repo:^github\.com/sourcegraph/sourcegraph$ Search(...)`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+Search%28...%29&patternType=structural).

Key thing to realize: when I add `@HEAD` to the repo, it disappears. This means some combination of the repo + branch/commit args that we send in searcher causes us to search a file we get from Zoekt and return results, but that result/file does not actually exist at that path for the branch/commit. So we should debug what we're sending/receiving when `@HEAD` added vs omitted.

### PR DESCRIPTION
**Pull request by camdencheek**
Fixes #17907 

Fixes an issue where results were being returned for branches other than the default branch when no revision was specified, causing `git ls-tree` errors in the frontend. 

The issue is that when we don't specify a revision, we were passing `branch == ""` to searcher, causing results to be returned from multiple branches (if there are multiple branches indexed for a repo). 

I think this solution should be fine (it's certainly better than what it was), but it's unclear to me if there is any situation where an empty revision means something other than just "search HEAD". It feels funny to not be using a specific revision name when we already queried Zoekt for indexed revisions, but every time I went down that path, it broke something else, so this feels like the safest option.

### CODE REVIEW

**Review by rvantonder — APPROVED** [review]
Fix looks good to me, thanks for investigating.

> it's unclear to me if there is any situation where an empty revision means something other than just "search HEAD".

@keegancsmith can you comment on ^?

---

Integration test would be nice, don't know if it's easy to replicate on our repo test set.

File: cmd/searcher/search/search_structural.go [hunk 1]
```diff
@@ -289,7 +289,11 @@ func structuralSearchWithZoekt(ctx context.Context, p *protocol.Request) (matche
 			Languages:                    p.Languages,
 		}
 
-	repoBranches := map[string][]string{string(p.Repo): {string(p.Branch)}}
+	branch := p.Branch
+	if branch == "" {
+		branch = "HEAD"
+	}
+	repoBranches := map[string][]string{string(p.Repo): {branch}}
```
Comment (rvantonder) [inline cmd/searcher/search/search_structural.go:296]:
```suggestion
	if p.Branch == "" {
		p.Branch = "HEAD"
	}
	repoBranches := map[string][]string{string(p.Repo): {p.Branch}}
```

I think this is fine?

Comment (camdencheek) [reply cmd/searcher/search/search_structural.go:296]: Seems fine to me. Doesn't look like `p` is used further down, so mutating it shouldn't hurt anything

**Comment by rvantonder** [comment]
I would not wait for @keegancsmith's response to merge--better to merge so we can nip this in the bud on prod :-)

**Comment by camdencheek** [comment]
> Integration test would be nice, don't know if it's easy to replicate on our repo test set.

It doesn't look like there are any repos with multiple branches indexed in the integration tests, so I don't think this issue would be easily tested there right now

**Comment by keegancsmith** [comment]
Empty means just HEAD.

**Pull request merged**